File: GW3/GWLogger/Backend/DataContext/Query/Tokens/TokenNumber.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenNumber : Token
    {
        const string allowedChar = "0123456789";

        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();
            if (parser.PeekChar() == '.' && allowedChar.IndexOf(parser.PeekChar(1)) != -1)
                return true;
            return allowedChar.IndexOf(parser.PeekChar()) != -1 && parser.PeekChar() != '.';
        }

        public override Token Extract(QueryParser parser)
        {
            var extracted = "";
            parser.SkipSpaces();

            while (parser.HasChar())
            {
                if (allowedChar.IndexOf(parser.PeekChar()) == -1 && parser.PeekChar() != '.')
                    break;
                extracted += parser.NextChar();
            }
            return new TokenNumber { Value = extracted };
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenNumber.ts
class TokenNumber extends Token {


    public proposals: SuggestionInterface[] = [];
    public tokenType: TokenType = TokenType.TokenNumber;
    private allowedChar: string = "0123456789";

    public canBeUsed(parser: QueryParser): boolean {
        parser.skipSpaces();
        if (parser.peekChar() == '.' && this.allowedChar.indexOf(parser.peekChar(1)) != -1)
            return true;
        return this.allowedChar.indexOf(parser.peekChar()) != -1 && parser.peekChar() != '.';
    }

    public extract(parser: QueryParser): Token {
        var extracted = "";
        parser.skipSpaces();

        while (parser.hasChar()) {
            if (this.allowedChar.indexOf(parser.peekChar()) == -1 && parser.peekChar() != '.')
                break;
            extracted += parser.nextChar();
        }
        var token = new TokenNumber();
        token.value = extracted;
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[] {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined') {
            switch (nextToken.tokenType) {
                case TokenType.TokenAnd:
                case TokenType.TokenOr:
                case TokenType.TokenWhere:
                    return [new TokenName(), new TokenOpenParenthesis()];
                case TokenType.TokenGroup:
                case TokenType.TokenOrder:
                    return [new TokenName()];
                case TokenType.TokenLimit:
                    return [new TokenNumber()];
                case TokenType.TokenCloseParenthesis:
                    return [
                        new TokenAnd(),
                        new TokenGroup(),
                        new TokenLimit(),
                        new TokenOr(),
                        new TokenOrder(),
                        new TokenWhere(),
                        new TokenName(),
                        new TokenCloseParenthesis()
                    ];
            }
        }
        return [];
    }
}